wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,587 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Region meta data
+
+use std::collections::{BTreeMap, HashMap};
+
+use common_types::{table::TableId, SequenceNumber};
+use common_util::define_result;
+use message_queue::Offset;
+use snafu::{Backtrace, OptionExt, Snafu};
+use tokio::sync::{Mutex, RwLock};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to update meta data after write of table:{}, msg:{}\nBacktrace:\n{}",
+        table_id,
+        msg,
+        backtrace
+    ))]
+    UpdateAfterWrite {
+        table_id: TableId,
+        msg: String,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Failed to mark deleted for table:{}, msg:{}\nBacktrace:\n{}",
+        table_id,
+        msg,
+        backtrace
+    ))]
+    MarkDeleted {
+        table_id: TableId,
+        msg: String,
+        backtrace: Backtrace,
+    },
+}
+define_result!(Error);
+
+// TODO: will be made use of later.
+#[allow(unused)]
+#[derive(Default, Debug)]
+pub struct RegionMeta {
+    inner: RwLock<RegionMetaInner>,
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+impl RegionMeta {
+    // TODO: Need to implement the init method using the [RegionMetaSnapshot] which
+    // will be persisted.
+
+    pub async fn prepare_for_table_write(&self, table_id: TableId) -> SequenceNumber {
+        {
+            let inner = self.inner.read().await;
+            if let Some(table_meta) = inner.table_metas.get(&table_id) {
+                return table_meta.get_meta_data().await.next_sequence_num;
+            }
+        }
+
+        // Double check is not needed, due to just one task will write the specific
+        // table.
+        let mut inner = self.inner.write().await;
+        assert!(inner
+            .table_metas
+            .insert(table_id, TableMeta::new(table_id))
+            .is_none(),
+            "now just support the thread model: one writer to one table, make no sense to occur race here");
+        // New table, so returned next sequence num is zero.
+        SequenceNumber::MIN
+    }
+
+    /// Update following meta data of table after each writing:
+    /// + mapping of start sequence number to start offset
+    /// + high watermark
+    /// + next sequence number
+    pub async fn update_after_table_write(
+        &self,
+        table_id: TableId,
+        write_offset_range: OffsetRange,
+        updated_num: u64,
```

Comment:
Is `updated_num` redundent? Because it can be computed by `(range.end - range.start + 1)`.

Reply:
> Is `updated_num` redundent? Because it can be computed by `(range.end - range.start + 1)`.

I think we should not assume that the returned offsets are continuous ?
Although it is indeed now.